scheduler/planner.go

```diff
@@ -0,0 +1,348 @@
+// in a Unit must be unique with regards to their ID.
+type Unit struct {
+	tasks       map[string]task.Task
+	cachedValue int64
+	id          string
+	distro      *distro.Distro
+}
+
+// MakeuUnit constructs a new unit, caching a reference to the distro
+// in the unit. It's valid to pass a nil here.
+func MakeUnit(d *distro.Distro) *Unit {
+	return &Unit{
+		distro: d,
+		tasks:  map[string]task.Task{},
+	}
+}
+
+// NewUnit constructs a new Unit container for a task.
+func NewUnit(t task.Task) *Unit {
+	u := MakeUnit(nil)
+	u.Add(t)
+	return u
+}
+
+// Export returns an unordered sequence of tasks from unit. All tasks
+// are unique.
+func (unit *Unit) Export() TaskList {
+	out := make(TaskList, 0, len(unit.tasks))
+
+	for _, t := range unit.tasks {
+		out = append(out, t)
+	}
+
+	return out
+}
+
+// Add caches a task in the unit.
+func (unit *Unit) Add(t task.Task) { unit.tasks[t.Id] = t }
+
+// SetDistro makes it possible to change/set the cached distro
+// reference in the unit; however, it is not possible to set a nil
+// distro.
+func (unit *Unit) SetDistro(d *distro.Distro) {
+	if d == nil || unit == nil {
+		return
+	}
+
+	unit.distro = d
+}
+
+func (unit *Unit) Keys() []string {
+	out := []string{}
+	for k := range unit.tasks {
+		out = append(out, k)
+	}
+	return out
+}
+
+// ID constructs a unique and hashed ID of all the tasks in the unit.
+func (unit *Unit) ID() string {
+	if unit.id != "" {
+		return unit.id
+	}
+
+	hash := sha1.New()
+	ids := make(sort.StringSlice, 0, len(unit.tasks))
+	for id := range unit.tasks {
+		ids = append(ids, id)
+	}
+	sort.Sort(ids)
+
+	for _, id := range ids {
+		_, _ = io.WriteString(hash, id)
+	}
+
+	unit.id = fmt.Sprintf("%x", hash.Sum(nil))
+	return unit.id
+}
+
+// RankValue returns a point value for the tasks in the unit that can
+// be used to compare units with eachother.
+//
+// Generally, higher point values are given to larger units and for
+// units that have been in the queue for longer, with longer expected
+// runtimes. The tasks priority act as a multiplying factor.
+func (unit *Unit) RankValue() int64 {
+	if unit.cachedValue > 0 {
+		return unit.cachedValue
+	}
+
+	var (
+		expectedRuntime time.Duration
+		timeInQueue     time.Duration
+		totalPriority   int64
+		numDeps         int64
+		inCommitQueue   bool
+		inPatch         bool
+	)
+
+	for _, t := range unit.tasks {
+		if t.Requester == evergreen.MergeTestRequester {
+			inCommitQueue = true
+		} else if evergreen.IsPatchRequester(t.Requester) {
+			inPatch = true
+		}
+
+		if !t.ScheduledTime.IsZero() {
+			timeInQueue += time.Since(t.ScheduledTime)
+		}
+
+		totalPriority += t.Priority
+		expectedRuntime += t.FetchExpectedDuration()
+		numDeps += int64(t.NumDependents)
+	}
+
+	num := int64(len(unit.tasks))
+	priority := 1 + (totalPriority / num)
+
+	if inCommitQueue {
+		priority += 100
+	}
+
+	if inPatch {
+		unit.cachedValue += unit.distro.GetPatchZipperFactor()
+	}
+
+	unit.cachedValue += num
```

Comment:
Include a comment with some examples and explain a bit about what guided this particular arithmetic.